parsing/pcoq.ml

```diff
@@ -432,10 +432,15 @@ type frozen_t =
 let freeze ~marshallable : frozen_t =
   (!grammar_stack, CLexer.get_keyword_state ())
 
+(* Physical equality here was missing a lot of same-entries, which
+   resulted in huge slowdown in Pcoq.freeze in some interactive users
+   *)
+let eq_grams x y = x = y
```

Comment:
Using structural equality on arbitrary data is actually a regression. You can get unexpected results that break the abstraction barrier (e.g. if we store sets / maps in there) or outright anomalies.

Reply:
It's not arbitrary however, isn't the type:
```
type frozen_t =
  (grammar_entry * GramState.t) list *
  CLexer.keyword_state
```

Reply:
Both GramState and grammar_entry can contain arbitary data.

Reply:
What if we force them to provide their own hashing functions, that way we can hash and compare? This is admittedly quite some work, but something worth doing if we want to compare data structures seriously.

Reply:
Indeed, this is the correct solution. We need to store equality functions when defining new grammar extension states.

Reply:
Best would be to refactor the parse state so we don't need to do this huge hack. This has been one of the top priorities IMHO engineering side for a while.

Reply:
But yes, I wanted to try a quick PR to see how this would affect CI / bench, let me fix this part.

Reply:
You can't really see the slowdown unless you go to CoqIDE etc...

Reply:
> How does the physical equality get lost?

I don't know, that's a good question. Ping me on Zulip if you are curious about how to reproduce this (as it a bit involved)

Reply:
Uh uh, I talked too fast, this seems pretty hard to fix, how I'm supposed to compare `Dyn` and `Store.S.t` types ?

Reply:
The situation with this Dyn types is pretty precarious IMO, @ppedrot, already it is state that any state in the grammar thing must be marhsallable, so if so, it should be OK to use the polymorphic equality.

IMHO we should forbid Marshall and have types declare the proper interfaces.

Reply:
Just register a map associating an equality function to a given tag when registering new grammar state. I don't see anything precarious in there, and you don't have to modify Dyn in any way.

Reply:
I need to compare Dyn values too, maybe I'm missing something.

The type of thing to compare is `grammar_entry * GramState.t` , there are directly derived from `Store` and Dyn

Reply:
Comparing Dyn tags is easy, there is a function for that. Using it, and assuming you registered a function for each tag you can then check the contents of the values. For Store it's harder because the API doesn't expose an equality function and it hides the underlying Dyn code but it's actually easy to implement.

Reply:
... and rereading the code, you actually don't need to compare the GramState.t values. Indeed, the invariant is that this state can be obtained by repeatedly applying the actions from the first component starting from the empty state, so in some sense they are redundant. Thus, in the end, you just have to tweak `Pcoq.create_entry_command` and `Pcoq.create_grammar_command` to take an additional equality function (you can put it in a record as `entry_extension` and `grammar_extension` if you want) and use that to compare the values when sharing the tails.

Reply:
> you can put it in a record as entry_extension and grammar_extension if you want

I'll try that out thanks, but I'm wary about these records being marshalled.